play-services-core/src/main/kotlin/com/google/android/gms/potokens/utils/PoTokenHelper.kt

```diff
@@ -0,0 +1,233 @@
+            return output
+        } catch (e: Throwable) {
+            e.printStackTrace()
+        }
+        return ByteArray(0)
+    }
+
+    private fun getDroidGuardResult(
+        context: Context?,
+        request: DroidGuardResultsRequest?,
+        map: Map<String, String>?
+    ): String? {
+        try {
+            val resultTask = DroidGuard.getClient(context).getResults("po-token-fast", map, request)
+            return Tasks.await(resultTask, 15, TimeUnit.SECONDS)
+        } catch (e: Throwable) {
+            e.printStackTrace()
+        }
+        return null
+    }
+
+    private fun getPoIntegrityToken(ctx: Context?, keySet: KeySet?): GetPoIntegrityTokenResponse? {
+        try {
+            val request = DroidGuardResultsRequest()
+            request.setTimeoutMillis(180000)
+            request.setOpenHandles(2)
+            request.setClientVersion(220221045)
+            keySet?.let {
+                request.bundle.putByteArray("po-fast-key", it.encode())
+            }
+            val tmpList = ArrayList<String>()
+            tmpList.add("po-fast-key")
+            request.bundle.putStringArrayList("extraKeysRetainedInFallback", tmpList)
+
+            val secureRandom = SecureRandom()
+            val randKeyBuf = ByteArray(0x20)
+            secureRandom.nextBytes(randKeyBuf)
+            val randKey = Base64.encodeToString(randKeyBuf, Base64.NO_WRAP)
+            val map: MutableMap<String, String> = HashMap()
+            map["b"] = randKey
+            val getPoIntegrityTokenRequestBuilder = GetPoIntegrityTokenRequest.Builder()
+            getDroidGuardResult(ctx, request, map)?.let {
+                getPoIntegrityTokenRequestBuilder.dgResult = it.encodeToByteArray().toByteString()
+            }
+            getPoIntegrityTokenRequestBuilder.dgRandKey = randKeyBuf.toByteString()
+            getPoIntegrityTokenRequestBuilder.mode = 1
+            val getPoIntegrityTokenRequest = getPoIntegrityTokenRequestBuilder.build()
+
+            return postPoTokenForGms(ctx, getPoIntegrityTokenRequest)
+        } catch (e: Throwable) {
+            e.printStackTrace()
+        }
+        return null
+    }
+
+    private fun postPoTokenForGms(
+        context: Context?,
+        request: GetPoIntegrityTokenRequest,
+    ): GetPoIntegrityTokenResponse? {
+        val future = RequestFuture.newFuture<GetPoIntegrityTokenResponse>()
+        Log.d(TAG,"postPoTokenForGms ")
+        Volley.newRequestQueue(context).add(object : Request<GetPoIntegrityTokenResponse>(
+            Method.POST, tokenUrl, future) {
+
+            override fun deliverResponse(response: GetPoIntegrityTokenResponse?) {
+                future.onResponse(response)
+            }
+
+            @Throws(AuthFailureError::class)
+            override fun getHeaders(): Map<String, String> {
+                return mapOf("User-Agent" to  "GmsCore/${Constants.GMS_VERSION_CODE} (${Build.DEVICE} ${Build.ID}); gzip")
```

Comment:
Shouldn't Build.DEVICE and Build.ID be the ones from selected device profile and not the real one?